### PR TITLE
71435 Fill tag table with area plus discipline description

### PR DESCRIPTION
**Pull request by janid1967**

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs [hunk 1]
```diff
@@ -48,43 +48,60 @@ public class CreateAreaTagCommandHandler : IRequestHandler<CreateAreaTagCommand,
...
-            var reqDefs =
-                await _requirementTypeRepository.GetRequirementDefinitionsByIdsAsync(reqDefIds);
-
             var project = await _projectRepository.GetByNameAsync(request.ProjectName);
             
             if (project == null)
             {
-                var mainProject = await _projectApiService.GetProject(_plantProvider.Plant, request.ProjectName);
-                if (mainProject == null)
+                project = await CreateProjectAsync(request.ProjectName);
```
Comment (TordJoranger) [inline src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: I think CreateProject is a bit missleading. We're not creating it, but fetching it from procosys (if i'm not mistaken?) and adding to the database. 
Something like FetchFromMain() or SyncWithMain() would be better I think.

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs [hunk 2]
```diff
@@ -48,43 +48,60 @@ public class CreateAreaTagCommandHandler : IRequestHandler<CreateAreaTagCommand,
...
-                project = new Project(_plantProvider.Plant, request.ProjectName, mainProject.Description);
-                _projectRepository.Add(project);
             }
 
-            var disciplines = await _disciplineApiService.GetDisciplines(_plantProvider.Plant);
-            var discipline = disciplines.SingleOrDefault(d => d.Code == request.DisciplineCode);
-            if (discipline == null)
+            var areaTagToAdd = await CreateAreaTagAsync(project, request);
+
+            if (!string.IsNullOrEmpty(request.AreaCode) && !await FillAreaDataAsync(areaTagToAdd, request.AreaCode))
```
Comment (TordJoranger) [inline src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: SetAreaCode() (or maybe something like SetAreaCodeSuccessully() so the bool return makes more sense).  
and do the nullceck inside the method

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs [hunk 3]
```diff
@@ -48,43 +48,60 @@ public class CreateAreaTagCommandHandler : IRequestHandler<CreateAreaTagCommand,
...
 
-                var area = areas.SingleOrDefault(a => a.Code == request.AreaCode);
+            return new SuccessResult<int>(areaTagToAdd.Id);
+        }
 
-                if (area == null)
-                {
-                    return new NotFoundResult<int>($"Area with code {request.AreaCode} not found");
-                }
+        private async Task<bool> FillDisciplineDataAsync(Tag tag, string disciplineCode)
```
Comment (TordJoranger) [inline src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: Same here, no reason to change from Set to Fill.

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateTag/CreateTagCommandHandler.cs [hunk 4]
```diff
@@ -64,39 +61,50 @@ public async Task<Result<List<int>>> Handle(CreateTagCommand request, Cancellati
...
+                McPkgNo = tagDetails.McPkgNo,
+                PurchaseOrderNo = tagDetails.PurchaseOrderNo,
+                Remark = request.Remark,
+                StorageArea = request.StorageArea,
+                TagFunctionCode = tagDetails.TagFunctionCode
+            };
+            tag.SetArea(tagDetails.AreaCode, tagDetails.AreaDescription);
+            tag.SetDiscipline(tagDetails.DisciplineCode, tagDetails.DisciplineDescription);
+            
+            project.AddTag(tag);
```
Comment (TordJoranger) [inline src/Equinor.Procosys.Preservation.Command/TagCommands/CreateTag/CreateTagCommandHandler.cs]: Take project.AddTag(tag) out of the method I think.

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs (hunk 1 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: Naah .. we create it in preservation context

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateTag/CreateTagCommandHandler.cs (hunk 4 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/TagCommands/CreateTag/CreateTagCommandHandler.cs]: ok. Will do the same in CreateTagCommandHandler

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs (hunk 1 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: Wont'fix

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs [hunk 5]
```diff
@@ -48,43 +48,65 @@ public class CreateAreaTagCommandHandler : IRequestHandler<CreateAreaTagCommand,
...
+            return new SuccessResult<int>(areaTagToAdd.Id);
+        }
 
-                if (area == null)
-                {
-                    return new NotFoundResult<int>($"Area with code {request.AreaCode} not found");
-                }
+        private async Task<bool> SetDisciplineDataSuccessfullyAsync(Tag tag, string disciplineCode)
+        {
+            var discipline = await _disciplineApiService.GetDiscipline(_plantProvider.Plant, disciplineCode);
```
Comment (TordJoranger) [inline src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: nullcheck on disiplinCode?

Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: No, required

Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: `public CreateAreaTagDtoValidator()
.
.
.
            RuleFor(x => x.DisciplineCode)
                .NotNull()
                .NotEmpty()
                .MaximumLength(Tag.DisciplineCodeLengthMax);`

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs [hunk 6]
```diff
@@ -48,43 +48,65 @@ public class CreateAreaTagCommandHandler : IRequestHandler<CreateAreaTagCommand,
...
             }
+            tag.SetDiscipline(disciplineCode, discipline.Description);
+            return true;
+        }
+
+        private async Task<bool> SetAreaDataSuccessfullyAsync(Tag tag, string areaCode)
+        {
+            if (string.IsNullOrEmpty(areaCode))
+            {
+                return true;
```
Comment (TordJoranger) [inline src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: bare dobbeltsjekker denne, Om requesten ikke indeholder areaCode, er det  OK, og derfor returnerer vi true?

Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/TagCommands/CreateAreaTag/CreateAreaTagCommandHandler.cs]: Ja

**Pull request merged**